### PR TITLE
Yamper CAN port 

### PR DESCRIPTION
**Pull request by ren-andy**
As title says, cleaned up Stephen's previous branch and finished the CAN library port. 

Some major changes made: 

- Changed clock freq to 42MHz, and the corresponding CAN baud rate 
- Modified makefile to build WLoopCAN when make is run in the root directory 
- Fixed compilation errors

### CODE REVIEW

File: STM32F042K4Tx/Core/Src/main.c [hunk 1]
```diff
@@ -45,44 +51,46 @@
...
 
-TIM_HandleTypeDef htim3;
-TIM_HandleTypeDef htim14;
+TIM_HandleTypeDef htim16;
 
 /* USER CODE BEGIN PV */
-float* velocity;
 uint8_t bytes_array[4];
 CAN_TxHeaderTypeDef TxHeader;
 extern uint32_t TxMailbox;
```
Comment (GuyBlumenthal) [inline STM32F042K4Tx/Core/Src/main.c]: You should be able to get rid of these

File: STM32F042K4Tx/Core/Src/main.c [hunk 2]
```diff
@@ -114,29 +121,49 @@ int main(void)
...
   MX_CAN_Init();
-  MX_TIM3_Init();
-  MX_TIM14_Init();
+  MX_TIM16_Init();
   /* USER CODE BEGIN 2 */
-  HAL_CAN_Start(&hcan);
-  HAL_TIM_Base_Start_IT(&htim14);
-  __HAL_TIM_SET_COUNTER(&htim14, 0);
+	HAL_CAN_Start(&hcan);
+	if(CANBus_init(&hcan, &htim16)!= HAL_OK)
```
Comment (GuyBlumenthal) [inline STM32F042K4Tx/Core/Src/main.c]: Why are you using the same timer for the can bus init and for the encoder?

File: STM32F042K4Tx/Core/Src/main.c [hunk 3]
```diff
@@ -208,89 +235,41 @@ static void MX_CAN_Init(void)
     Error_Handler();
   }
   /* USER CODE BEGIN CAN_Init 2 */
-  TxHeader.StdId = 0x1F;
-  TxHeader.ExtId = 0;
+	TxHeader.StdId = 0x1F;
+	TxHeader.ExtId = 0;
```
Comment (GuyBlumenthal) [inline STM32F042K4Tx/Core/Src/main.c]: You don't need this anymore

File: STM32F042K4Tx/Core/Src/main.c [hunk 4]
```diff
@@ -311,19 +290,42 @@ static void MX_GPIO_Init(void)
...
+  HAL_NVIC_EnableIRQ(EXTI4_15_IRQn);
+
 }
 
 /* USER CODE BEGIN 4 */
-void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim){
-	HAL_GPIO_TogglePin(GPIOB,GPIO_PIN_7);
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+	HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_7);
+}
```
Comment (GuyBlumenthal) [inline STM32F042K4Tx/Core/Src/main.c]: You need to call the Wloop can library timer ISR here

File: makefile [hunk 5]
```diff
@@ -137,7 +139,7 @@ LIBDIR =
 LDFLAGS = $(MCU) -specs=nano.specs -T$(LDSCRIPT) $(LIBDIR) $(LIBS) -Wl,-Map=$(BUILD_DIR)/$(TARGET).map,--cref -Wl,--gc-sections
 
 # default action: build all
-all: $(BUILD_DIR)/$(TARGET).elf $(BUILD_DIR)/$(TARGET).hex $(BUILD_DIR)/$(TARGET).bin
+all: can_library $(BUILD_DIR)/$(TARGET).elf $(BUILD_DIR)/$(TARGET).hex $(BUILD_DIR)/$(TARGET).bin
```
Comment (GuyBlumenthal) [inline makefile]: can you change the name of the target from `can_library` to `libs` so we can include the util library in it in the future

File: STM32F042K4Tx/Core/Src/main.c (hunk 2 above)
Comment (ren-andy) [reply STM32F042K4Tx/Core/Src/main.c]: This is a good question... I'm not exactly sure why this is, but I'll separate the CAN and encoder into two different timers.

File: STM32F042K4Tx/Core/Src/main.c (hunk 3 above)
Comment (ren-andy) [reply STM32F042K4Tx/Core/Src/main.c]: - [x] remove unnecessary code

File: STM32F042K4Tx/Core/Src/main.c (hunk 4 above)
Comment (ren-andy) [reply STM32F042K4Tx/Core/Src/main.c]: - [x] add WLoopCAN ISR timer call here

File: makefile (hunk 5 above)
Comment (ren-andy) [reply makefile]: - [x] change target name to libs for extensibility

File: STM32F042K4Tx/Core/Src/main.c (hunk 1 above)
Comment (ren-andy) [reply STM32F042K4Tx/Core/Src/main.c]: - [x] remove unused CAN header defs

File: STM32F042K4Tx/Core/Src/main.c [hunk 6]
```diff
@@ -341,8 +343,17 @@ static void MX_GPIO_Init(void)
...
 void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+  /* 
+   * tim14 should call interrupt every 200 ms
+   * freq = clock freq / (prescalar + 1) * (counter period + 1) * (repetition counter + 1)
+   *      = 42 MHz/(99 + 1)(41999+1)(1)
+   *      = 42 MHz/(100)(42000)
+   *      = 10 Hz 
+   * period = 1/10 s = 100 ms
+   */ 
+
```
Comment (GuyBlumenthal) [inline STM32F042K4Tx/Core/Src/main.c]: Ok, so you are calculating tim14 as your WLoop CAN timer but you are using tim16 for the can init and tim14 for the calculating encoder stuff?

Comment (ren-andy) [reply STM32F042K4Tx/Core/Src/main.c]: Fixed. TIM14 should now just be a counter with no interrupt - used by the ring encoder; and TIM16 now has an interrupt which is called every 500ms - used by WLoopCAN

**Review by GuyBlumenthal — APPROVED** [review]

**Pull request merged**